cis194/week1/daniel-deng/src/CIS194/Homework02/Exercise01.hs

```diff
@@ -0,0 +1,34 @@
+module CIS194.Homework02.Exercise01 (parseMessage, parse) where
+
+import CIS194.Homework02.Log
+
+import Control.Monad.Zip
+import Text.Read
+
+maybeInt :: String -> Maybe Int
+maybeInt = readMaybe
+
+info :: String -> [String] -> LogMessage
+info ts body = case maybeInt ts of
+                 Just timestamp -> LogMessage Info timestamp (unwords body)
+                 Nothing        -> Unknown (unwords $ ["I", ts] ++ body)
+
+warn :: String -> [String] -> LogMessage
+warn ts body = case maybeInt ts of
+                 Just timestamp -> LogMessage Warning timestamp (unwords body)
+                 Nothing        -> Unknown (unwords $ ["W", ts] ++ body)
+
+err :: String -> String -> [String] -> LogMessage
+err errCode ts body = case maybeInt errCode `mzip` maybeInt ts of
+                        Just (code, timestamp) -> LogMessage (Error code) timestamp (unwords body)
+                        Nothing                -> Unknown (unwords $ ["E", errCode, ts] ++ body)
+
+parseMessage :: String -> LogMessage
+parseMessage logLine = case words logLine of
```

Comment:
As a advanced technic view patterns extension may simplify this file significantly https://gitlab.haskell.org/ghc/ghc/wikis/view-patterns-alternative
Have a look at the above link and see if it can benefit you.

Reply:
```Haskell
parseMessage ( words -> "E"
             : ( readMaybe -> Just ( lv :: Integer ) )
             : rest ) = ... lv .... -- lv can be used here directly
parseMessage ....
```

Reply:
Thanks for the tip @shineli1984 ! That gives me much motivation to look into language extensions 🤣

Reply:
https://github.com/fp-works/2019-winter-Haskell-school/pull/18/commits/65f2b30f5aa172c296a2da0bfd7ff3b1e0295dff @shineli1984